lib/matplotlib/axes/_axes.py

```diff
@@ -3504,7 +3504,9 @@ def apply_mask(arrays, mask): return [array[mask] for array in arrays]
                     f"'{dep_axis}err' (shape: {np.shape(err)}) must be a "
                     f"scalar or a 1D or (2, n) array-like whose shape matches "
                     f"'{dep_axis}' (shape: {np.shape(dep)})") from None
-            if np.any(err < -err):  # like err<0, but also works for timedelta.
```

Comment:
Could you check for the nan in the conditional?
`(err < -err) & ~np.isnan(err)`

Reply:
`isnan` doesn't work for `dtype=object` https://stackoverflow.com/questions/36198118/np-isnan-on-arrays-of-dtype-object and casting to float will break `timedelta`.

There may be a better way to write it, but that is the reason for the current approach.

Reply:
Maybe `& err != err`?

Reply:
Ah, there will be runtime warnings that will fail the tests.

```
err = np.array([np.nan, -0.1], dtype=object)
np.any(err < -err)
```
gives
```
/tmp/ipykernel_32424/1840248036.py:1: RuntimeWarning: invalid value encountered in less
  np.any(err < -err)
```
independent of what is added (unless something like `err != err` is added before, but then it will not test properly).

Reply:
(It took me quite some time to get a solution that worked... Not that it may be a good indication, but at least I failed with quite a few alternatives...)

Reply:
Shouldn't timedelta have been converted to float before now?  I think process_unit_info should have converted them to floats?  If not, we should fix that...

Reply:
Maybe it should, but it isn't.

Doing `err = err.astype(float)` before the old if-clause fails:
https://github.com/matplotlib/matplotlib/blob/953790680ac667adb869e1f1cba2f2b534cc6f0b/lib/matplotlib/tests/test_axes.py#L3665-L3671

I really have no idea how to deal with units, so that will be a different fix then.

Reply:
I'm a bit confused here - how do we make a bar with timedeltas on it?  I actually get that to fail as is.

Reply:
Changing the above example to `days=10` gives:

![image](https://user-images.githubusercontent.com/8114497/167383557-e16b7dd0-6b96-4542-87ba-f2280cd63411.png)

Reply:
This is definitely another situation where things have been made _very_ complicated by carrying units around for too long.  We don't even know if `err` is negative or positive until its units have been converted, so I'm not sure that this check should be here?

Reply:
I do not know how unit handling works at all, but I was thinking about the sign issue and I cannot get my head around it. When would units affect the sign? I may miss something obvious here though.

The only thing I can remotely think about is temperature where 3 degrees F is -16 degrees C, but that is really more a consequence of bad maths (and the reason that `timedelta` is used, rather than an absolute date/time for `yerr` in the example).

Then one can of course discuss if the sign is important or if one just should to `abs` at some stage (if that is even required). The errors are assumed symmetrical here anyway.

Reply:
> When would units affect the sign?

Units can convert anything to anything - we can't tell if they affect sign or not, even for deltas of various sorts.  The point is that at this point, if we have not converted to numpy, then the negative check is not guaranteed to work properly.

Reply:
But it is numpy? Although with `dtype=object`. Or am I missing something?

Either way, I am not really sure how to continue here. The PR solves #22910 without introducing any new issues (as far as I can tell, all existing tests pass anyway). Although it may be that lots of other related things also should be fixed.

It would really help to get some examples of things that doesn't work though.

Reply:
> But it is numpy? Although with dtype=object. Or am I missing something?

Yeah, but dtype object isn't guaranteed to do anything under mathematical operations (of particular relevance `<` is not guaranteed to work).  Its basically just a list.  

> The PR solves https://github.com/matplotlib/matplotlib/issues/22910 without introducing any new issues (as far as I can tell, all existing tests pass anyway). Although it may be that lots of other related things also should be fixed.

Fair, maybe this is a tangent of sorts, but I think we need to fix the underlying problems rather than add more bandaids on top.  

> It would really help to get some examples of things that doesn't work though.

I'm 99% sure that if you passed a unit-full pint object in as `err` this would probably fail.  Maybe it failed before as well?

I'm not blocking on this, but just pointing out that we still have substantial work to do on cleaning up the engineering here.

Reply:
I agree that it is bandaid for sure.

Thanks! Wasn't aware of pint (although I've seen it mentioned). This is what happens on both `main` and this when using `errorbar`:
```
/local/data1/matplotlib/lib/matplotlib/axes/_axes.py:3329: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  y = np.asarray(y, dtype=object)
/local/data1/matplotlib/lib/matplotlib/axes/_axes.py:3374: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  return np.asarray(err, dtype=object)
```
![image](https://user-images.githubusercontent.com/8114497/168816484-3146b833-8b23-4c60-b9aa-04fa4809f938.png)

and when using `bar`:
```
/local/data1/miniconda3/lib/python3.8/site-packages/numpy/lib/stride_tricks.py:536: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  args = [np.array(_m, copy=False, subok=subok) for _m in args]
Traceback (most recent call last):

  File "/local/data1/miniconda3/lib/python3.8/site-packages/pint/quantity.py", line 901, in __float__
    raise DimensionalityError(self._units, "dimensionless")

DimensionalityError: Cannot convert from 'meter' to 'dimensionless'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):

  File "/tmp/ipykernel_13475/2206117657.py", line 1, in <cell line: 1>
    plt.bar([1, 2], base, yerr=err)

  File "/local/data1/matplotlib/lib/matplotlib/pyplot.py", line 2392, in bar
    return gca().bar(

  File "/local/data1/matplotlib/lib/matplotlib/__init__.py", line 1412, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)

  File "/local/data1/matplotlib/lib/matplotlib/axes/_axes.py", line 2369, in bar
    x, height, width, y, linewidth, hatch = np.broadcast_arrays(

  File "<__array_function__ internals>", line 5, in broadcast_arrays

  File "/local/data1/miniconda3/lib/python3.8/site-packages/numpy/lib/stride_tricks.py", line 536, in broadcast_arrays
    args = [np.array(_m, copy=False, subok=subok) for _m in args]

  File "/local/data1/miniconda3/lib/python3.8/site-packages/numpy/lib/stride_tricks.py", line 536, in <listcomp>
    args = [np.array(_m, copy=False, subok=subok) for _m in args]

ValueError: setting an array element with a sequence.
```
(empty plot).

Reply:
Fair, so the point is you are definitelynot making it worse here, but the problem still remains.

Reply:
I guess that a quick bandaid for `bar` and units is to use this approach there as well:
https://github.com/matplotlib/matplotlib/blob/f25c2d025fa8f70d6688e7acf4b85e4eb7590299/lib/matplotlib/axes/_axes.py#L3357-L3361

Reply:
No, that is what I'm fundamentally objecting to - just making an object array doesn't make any sense!